asllib/ASLTypingReference.mld

```diff
@@ -722,124 +722,261 @@ of the following applies:
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a while statement with expression [e1] and statement [s1];
+    - [t, e2] is the result of annotating [e1] in [env];
+    - [t] type-satisfies [t_bool];
+    - [s2] is the result of annotating [s1] in [env];
+    - [new_s] is a while statement with expression [e2] and statement [s2];
+    - [new_env] is [env].
 
     {4 Example}
 
   {3 TypingRule.SRepeat}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a repeat statement with expression [e1] and statement [s1];
+    - [s2] is the result of annotating [s1] in [env];
+    - [t, e2] is the result of annotating [e1] in [env];
+    - [t] type-satisfies [t_bool];
+    - [new_s] is a repeat statement with expression [e2] and statement [s2];
+    - [new_env] is [env].
 
     {4 Example}
 
   {3 TypingRule.SFor}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a for statement with index [id], direction [dir], two expressions
+      [e1] and [e2] and a statement [s'];
+    - [t1,e1'] is the result of annotating [e1] in [env];
+    - [t2,e2'] is the result of annotating [e2] in [env];
+    - [t1] has the structure of an integer type;
+    - [t2] has the structure of an integer type;
+    - [ty] is ;
+    - [s''] is the result of annotating [s'] in [env];
+    - [new_s] is a for statement with index [id], direction [dir], two expressions
+      [e1'] and [e2'] and statement [s''];
+    - [new_env] is [env].
 
     {4 Example}
 
-  {3 TypingRule.SThrowNone}
+  {3 TypingRule.SDeclSome}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a declaration with [ldk], [ldi] and an optional expression [e_opt];
+    - [e_opt] gives some expression [e];
+    - [t_e,e'] is the result of annotating [e] in [env];
+    - [env', ldi'] is the result of declaring the local constants of [ldi] in [env];
+    - [new_s] is a declaration with [ldk], [ldi'] and an expression [e'];
+    - [new_env] is [env].
 
     {4 Example}
 
-  {3 TypingRule.SThrowSome}
+  {3 TypingRule.SDeclNone}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a declaration statement with [ldk], [ldi] and an optional expression [e_opt];
+    - [e_opt] gives no expression;
+    - [env', s'] is the result of annotating uninitialised local declarations [ldi] in [env];
+    - [new_s] is [s'];
+    - [new_env] is [env].
 
     {4 Example}
-    <minimal example>
 
-  {3 TypingRule.STry}
+  {3 TypingRule.SThrowNone}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a throw statement with no expression;
+    - [new_s] is [s];
+    - [new_env] is [env].
 
     {4 Example}
 
-  {3 TypingRule.SDeclSome}
+  {3 TypingRule.SThrowSome}
 
     {4 Rule}
+    All of the following applies:
+    - [s] is a throw statement with expression [e];
+    - [t_e,e'] is the result of annotating [e] in [env];
+    - [t_e] has the structure of an exception type;
+    - [new_s] is a throw statement with expression [e'] and type [t_e];
```

Comment:
This is actually not syntax, so we might want to specify somewhere that we are adding some typing information in the AST.
Maybe `AST.mli` deserves an intro on that point?